Adding ChiselEnum Documentation Entry (#1795)
chipsalliance/chisel3: docs/src/explanations/chisel-enum.md

# ChiselEnum

> The ChiselEnum type is an easy way to get away from value encodings and reduce the chance of error when encoding muxes, opcodes, and functional unit operations. As opposed to regular Enums, ChiselEnums allow for function names to be explicitly passed through Bundles like IO. 

Reviewer: Thanks for writing this!

what is a "regular Enum"? Do you mean `scala.Enum`, or something else?

Also, not sure what you mean by "allow for function names". What is a "function name"?

Author: By regular Enum I mean the Chisel enum. I can clarify that. 
By function name I mean like a opcode being passed into an ALU or CMP.

## Importing the ChiselEnum module

```scala mdoc:silent
import chisel3._
import chisel3.experimental.ChiselEnum
```

## Functionality and Examples

Below we see ChiselEnum being used as mux select for a RISC-V core. While wrapping the object in a package is not required, it is highly recommended as it allows for the type to be used in multiple files more easily. 

```scala mdoc:silent
package CPUTypes {
    object AluMux1Sel extends ChiselEnum {
        val rs1out, pcout = Value
        /**
            "rs1out" -> 0.U,
            "pcout"  -> 1.U
        */
    }
}
```

Here we see a mux using the AluMux1Sel to select between different inputs. 

```scala mdoc:silent
import CPUTypes.AluMux1Sel
import CPUTypes.AluMux1Sel._

class AluMux1File(val dl_size: Int) extends Module {
    val io = IO(new Bundle {
        val aluMux1Sel =  Input( AluMux1Sel() )
        val rs1Out     =  Input(Bits(dl_size.W))
        val pcOut      =  Input(Bits(dl_size.W))
        val aluMux1Out = Output(Bits(dl_size.W))
    })

    // Default value for aluMux1Out

>     io.aluMux1Out := 0.U

Reviewer: shouldn't the default be one of the `ChiselEnum` types?

Author: It can be a default value depends on whether logic is being simplified or error checked. Something like `DontCare` or `0.U` can be used for checking that unknown selects are used. I believe that this is a matter of implementation.

Reviewer: Oh, I misread the example, the type is on the select not the data.

Author: Correct!

    switch (io.aluMux1Sel) {
        is (rs1out) {
            io.aluMux1Out  := io.rs1Out
        }
        is (pcout) {
            io.aluMux1Out  := io.pcOut
        }
    }
}
```
```scala mdoc:verilog
import 
ChiselStage.emitVerilog.emitVerilog(new AluMux1File(dl_size = 32) )
```

> ChiselEnum also allows for the user to define variables by passing in the value shown below. Note that the value must be increasing.

Reviewer: What happens if the value isn't increasing? Is it a runtime error?

Author: An exception is thrown when Chisel emits Verilog.
> chisel3.internal.ChiselException: Exception thrown when elaborating ChiselGeneratorAnnotation

```scala mdoc:silent
object Opcode extends ChiselEnum {
    val load  = Value(0x03.U) // i load  -> 000_0011
    val imm   = Value(0x13.U) // i imm   -> 001_0011
    val auipc = Value(0x17.U) // u auipc -> 001_0111
    val store = Value(0x23.U) // s store -> 010_0011
    val reg   = Value(0x33.U) // r reg   -> 011_0011
    val lui   = Value(0x37.U) // u lui   -> 011_0111
    val br    = Value(0x63.U) // b br    -> 110_0011
    val jalr  = Value(0x67.U) // i jalr  -> 110_0111
    val jal   = Value(0x6F.U) // j jal   -> 110_1111
}
```

The user can 'jump' to a value and continue incrementing by passing a start point then using a regular Value assignment. 

```scala mdoc:silent
object BranchFunct3 extends ChiselEnum {
    val beq, bne = Value
    val blt = Value(4.U)
    val bge, bltu, bgeu = Value
    /**
        "beq"  -> 0.U,
        "bne"  -> 1.U,
        "blt"  -> 4.U,
        "bge"  -> 5.U,
        "bltu" -> 6.U,
        "bgeu" -> 7.U
    */
}
```

## Testing

> When testing your modules, the `.Type` and `.litValue` attributes allow for the the objects to be passed as parameters and for the value to be converted to BigInt type. Note that BigInts cannot be casted to Int with `.asInstanceOf[Int]`, they use their own methods like `toInt`. Please review the [scala.math.BigInt](https://www.scala-lang.org/api/2.12.5/scala/math/BigInt.html) page for more details!

Reviewer: the example below is a nice example of `litValue`, can you show an example of `.Type` usage?

Also, does `sel.litValue == 0.U` also work? It seems more intuitive... actually it seems way more intuitive to actually use the Enum values in such a comparison. The below seems like a pattern that should be discouraged

Author: .Type is used in line 100. In the testbench, the .Type function is used to pass the ChiselEnum as a parameter since we do not want the literal value!

Author: `sel.litValue == 0.U` does not work since .litValue returns a BigInt and not a UInt. This block is pulled out of my testbench which tests that the ChiselEnum type is enumerating correctly and that all values match as expected.

Reviewer: gotcha. I can't resolve this comment for some reason but you've answered my question

```scala mdoc:silent
def expectedSel(sel: AluMux1Sel.Type): Boolean = sel match {
  case AluMux1Sel.rs1out => (sel.litValue == 0)
  case AluMux1Sel.pcout  => (sel.litValue == 1)
  case _                 => false
}
```

The ChiselEnum type also has methods `.all` and `.getWidth` where all returns all of the enum instances and getWidth returns the width of the hardware type.

## Workarounds

As of 2/26/2021, the width of the values is always infered so to get around this, just add an extra state that forces the width that is desired. 

```scala mdoc:silent
object StoreFunct3 extends ChiselEnum {
    val sb, sh, sw = Value
    val ukn = Value(7.U)
    /**
        "sb" -> 0.U,
        "sh" -> 1.U,
        "sw" -> 2.U
    */
}
```

> Additionally I don't believe that signed values are supported so if you want the value signed, you must cast the UInt with `.asSInt`.

Reviewer: Before this PR is merged let's get rid of the `I don't believe` and just make sure what is written is correct.

# ChiselEnum

The ChiselEnum type can be used to reduce the chance of error when encoding mux selectors, opcodes, and functional unit operations. In contrast with`Chisel.util.Enum`, `ChiselEnum` are subclasses of `Data`, which means that they can be used to define fields in `Bundle`s, including in `IO`s.

## Functionality and Examples

```scala mdoc
// Imports used in the following examples
import chisel3._
import chisel3.util._
import chisel3.stage.ChiselStage

> import chisel3.experimental.ChiselEnum

Reviewer: ```suggestion
import chisel3.experimental.ChiselEnum
```

```

Reviewer: sigh. i swear i am suggesting the additional backticks!

Author: I pushed a new commit with backticks

# ChiselEnum

> The ChiselEnum type is an easy way to get away from value encodings and reduce the chance of error when encoding muxes, opcodes, and functional unit operations. As opposed to the Chisel type Enums, ChiselEnums allow for names to be explicitly passed through Bundles like IO. 

Reviewer: ```suggestion
The ChiselEnum type can be used to reduce the chance of error when encoding mux selectors, opcodes, and functional unit operations. In contrast with`Chisel.Enum`, `ChiselEnum` can be included in `Bundle`s.
```

I still don't know what you mean by "names to be explicitly passed" or "through Bundles like IO". Even from an english language level, I don't know if "like IO" means "Bundles which are similar to IO" or "similar to the way IO can be passed through Bundles" Is the above correct?

Author: Well I wanted to explicitly mention `IO` because it is the most prevalent use of `Bundle`s. When I was first trying to use `Enum` for my mux selects, I couldn't pass the `Enum` type through the `IO` of my modules. This is only possible with `ChiselEnum`s which define a `Data` type that can be passed through Bundles and IO. Since `IO` takes in a `Bundle` object, I was trying to make it clear that `ChiselEnum`s can be passed through `Bundle`s like those used in `IO` while `Enum` objects cannot

Reviewer: thanks for the explanation! What do you think of my suggested change above? (Feel free to use any/all/none of it)

Author: It looks good. The only thing that I would change is that `Enum` is in the `Chisel.util` package, not in `Chisel`. I don't know if you were trying to reference the package or not.

Reviewer: yeah I was trying to reference the package to be explicit what we meant by "Enum". I'll fix it to say util

> ## Importing the ChiselEnum module
> 
> ```scala mdoc
> import chisel3._
> import chisel3.util._
> import chisel3.stage.ChiselStage
> import chisel3.experimental.ChiselEnum
> 
> ```
> 
> ## Functionality and Examples
> 

Reviewer: ```suggestion

## Functionality and Examples

```scala mdoc
// Imports used in the following examples
import chisel3._
import chisel3.util._
import chisel3.stage.ChiselStage
import chisel3.experimental.ChiselEnum
```

```

Author: I had the import in separate ticks from the example. You removed the end ticks from "Importing the ChiselEnum module" but not the start ticks of the next block

Reviewer: yeah but i think that is an artifact of github's suggest changes API... anyway, consider this just as a suggestion

Below we see ChiselEnum being used as mux select for a RISC-V core. While wrapping the object in a package is not required, it is highly recommended as it allows for the type to be used in multiple files more easily. 

```scala mdoc
// package CPUTypes {
    object AluMux1Sel extends ChiselEnum {

>         val rs1out, pcout = Value
>         /**
>             "rs1out" -> 0.U,
>             "pcout"  -> 1.U
>         */

Reviewer: ```suggestion
        val selectRS1, selectPC = Value
        /**
            "selectRS1" -> 0.U,
            "selectPC"  -> 1.U
        */
```

for the purposes of a clear example, can I suggest more generic names above?

    }
// }
```

Here we see a mux using the AluMux1Sel to select between different inputs. 

```scala mdoc
import AluMux1Sel._

class AluMux1File extends Module {

>     val io = IO(new Bundle {

Reviewer: i think this is the reason your thing is failing CI, there is a bug that shows up in `mdoc` + naming plugin +` IO( new Bundle {...})`

You can try workaround this by defining your bundle type outside of the IO call:

```
class ALUMuxIO extends Bundle {
...
}
...

val io = IO(new ALUMuxIO)
```

Author: Oh that's a pretty nasty issue. I separated the bundle class out and looks like it is passing the ci now. Thanks!

        val aluMux1Sel =  Input( AluMux1Sel() )
        val rs1Out     =  Input(Bits(32.W))
        val pcOut      =  Input(Bits(32.W))
        val aluMux1Out = Output(Bits(32.W))
    })

    // Default value for aluMux1Out
    io.aluMux1Out := 0.U

    switch (io.aluMux1Sel) {
        is (rs1out) {
            io.aluMux1Out  := io.rs1Out
        }
        is (pcout) {
            io.aluMux1Out  := io.pcOut
        }
    }
}
```
```scala mdoc:verilog
ChiselStage.emitVerilog(new AluMux1File )
```

ChiselEnum also allows for the user to define variables by passing in the value shown below. Note that the value must be increasing or else 
 > chisel3.internal.ChiselException: Exception thrown when elaborating ChiselGeneratorAnnotation

> is thrown during Verliog generation.

Reviewer: something weird here in the formatting. You can actually make the `mdoc` just display the thrown error by doing the bad thing in the code

Author: I added space around the block quote for now. I'm worried that the ci will start going crazy if I intentionally throw errors in the mdoc

Reviewer 2: For an expected exception, I think `scala mdoc:crash` is what you want. There's also `scala mdoc:fail`. Lots of documentation on this [here](https://scalameta.org/mdoc/docs/modifiers.html).

    val jalr  = Value(0x67.U) // i jalr  -> 110_0111
    val jal   = Value(0x6F.U) // j jal   -> 110_1111
}
```

The user can 'jump' to a value and continue incrementing by passing a start point then using a regular Value assignment. 

```scala 
object BranchFunct3 extends ChiselEnum {
    val beq, bne = Value
    val blt = Value(4.U)
    val bge, bltu, bgeu = Value
    /**
        "beq"  -> 0.U,
        "bne"  -> 1.U,
        "blt"  -> 4.U,
        "bge"  -> 5.U,
        "bltu" -> 6.U,
        "bgeu" -> 7.U
    */
}
```

## Testing

When testing your modules, the `.Type` and `.litValue` attributes allow for the the objects to be passed as parameters and for the value to be converted to BigInt type. Note that BigInts cannot be casted to Int with `.asInstanceOf[Int]`, they use their own methods like `toInt`. Please review the [scala.math.BigInt](https://www.scala-lang.org/api/2.12.5/scala/math/BigInt.html) page for more details!

```scala 
def expectedSel(sel: AluMux1Sel.Type): Boolean = sel match {
  case AluMux1Sel.rs1out => (sel.litValue == 0)
  case AluMux1Sel.pcout  => (sel.litValue == 1)
  case _                 => false
}
```

> The ChiselEnum type also has methods `.all` and `.getWidth` where all returns all of the enum instances and getWidth returns the width of the hardware type.

Reviewer: ```suggestion
The ChiselEnum type also has methods `.all` and `.getWidth` where `all` returns all of the enum instances and `getWidth` returns the width of the hardware type.
```

# ChiselEnum

The ChiselEnum type is an easy way to get away from value encodings and reduce the chance of error when encoding muxes, opcodes, and functional unit operations. As opposed to the Chisel type Enums, ChiselEnums allow for names to be explicitly passed through Bundles like IO. 

## Importing the ChiselEnum module

> ```scala mdoc

Reviewer: can you break these up or make some of these invisible/silent since `ChiselStage` is not relevant to to "Importing the Chisel Enum module" ?

Reviewer: I'm not sure the general pattern in these docs if things that are only relevant for the examples are hidden or if the code examples are all supposed to work right out of the box

Author: In referencing https://raw.githubusercontent.com/chipsalliance/chisel3/master/docs/src/explanations/naming.md they also had ChiselStage shown

Reviewer: yeah, gotcha. Then let's follow that pattern a bit more ... I'll suggest a change

Author: I also believe this helps people who want to copy what is on screen and get it to run locally on their machine.

```scala mdoc
// package CPUTypes {
    object AluMux1Sel extends ChiselEnum {
        val selectRS1, selectPC = Value
        /**
            "selectRS1" -> 0.U,
            "selectPC"  -> 1.U
        */
    }
// }
```

Here we see a mux using the AluMux1Sel to select between different inputs. 

```scala mdoc
import AluMux1Sel._

class AluMux1Bundle extends Bundle {
        val aluMux1Sel =  Input( AluMux1Sel() )
        val rs1Out     =  Input(Bits(32.W))
        val pcOut      =  Input(Bits(32.W))
        val aluMux1Out = Output(Bits(32.W))
}

class AluMux1File extends Module {
    val io = IO(new AluMux1Bundle)

    // Default value for aluMux1Out
    io.aluMux1Out := 0.U

    switch (io.aluMux1Sel) {
        is (selectRS1) {
            io.aluMux1Out  := io.rs1Out
        }
        is (selectPC) {
            io.aluMux1Out  := io.pcOut
        }
    }
}
```
```scala mdoc:verilog
ChiselStage.emitVerilog(new AluMux1File )
```

ChiselEnum also allows for the user to define variables by passing in the value shown below. Note that the value must be increasing or else 
 > chisel3.internal.ChiselException: Exception thrown when elaborating ChiselGeneratorAnnotation

> is thrown during Verliog generation.

Reviewer: ```suggestion
is thrown during Verilog generation.
```

```scala 
object BranchFunct3 extends ChiselEnum {
    val beq, bne = Value
    val blt = Value(4.U)
    val bge, bltu, bgeu = Value
    /**
        "beq"  -> 0.U,
        "bne"  -> 1.U,
        "blt"  -> 4.U,
        "bge"  -> 5.U,
        "bltu" -> 6.U,
        "bgeu" -> 7.U
    */
}
```

## Testing

When testing your modules, the `.Type` and `.litValue` attributes allow for the the objects to be passed as parameters and for the value to be converted to BigInt type. Note that BigInts cannot be casted to Int with `.asInstanceOf[Int]`, they use their own methods like `toInt`. Please review the [scala.math.BigInt](https://www.scala-lang.org/api/2.12.5/scala/math/BigInt.html) page for more details!

```scala 
def expectedSel(sel: AluMux1Sel.Type): Boolean = sel match {
  case AluMux1Sel.selectRS1 => (sel.litValue == 0)
  case AluMux1Sel.selectPC  => (sel.litValue == 1)
  case _                 => false
}
```

The ChiselEnum type also has methods `.all` and `.getWidth` where `all` returns all of the enum instances and `getWidth` returns the width of the hardware type.

## Workarounds

> As of 2/26/2021, the width of the values is always infered so to get around this, just add an extra state that forces the width that is desired. 

Reviewer: ```suggestion
As of 2/26/2021, the width of the values is always inferred. To work around this, you can add an extra `Value` that forces the width that is desired. This is shown in the example below, where we add a field `ukn` to force the width to be 3 bits wide: 
```

# ChiselEnum

The ChiselEnum type can be used to reduce the chance of error when encoding mux selectors, opcodes, and functional unit operations. In contrast with`Chisel.util.Enum`, `ChiselEnum` are subclasses of `Data`, which means that they can be used to define fields in `Bundle`s, including in `IO`s.

## Functionality and Examples

```scala mdoc
// Imports used in the following examples
import chisel3._
import chisel3.util._
import chisel3.stage.ChiselStage
import chisel3.experimental.ChiselEnum
```

Below we see ChiselEnum being used as mux select for a RISC-V core. While wrapping the object in a package is not required, it is highly recommended as it allows for the type to be used in multiple files more easily. 

```scala mdoc
// package CPUTypes {
    object AluMux1Sel extends ChiselEnum {
        val selectRS1, selectPC = Value
        /**
            "selectRS1" -> 0.U,
            "selectPC"  -> 1.U
        */
    }
// }

> ```

Reviewer 2: To narrowly document `ChiselEnum`, it's sufficient to ignore suggestions of how a user should organize/package their code.

That said, if you want to show this as living under `CPUTypes`, I think that's reasonable. Because Scala seemingly treats objects and packages as basically the same, you can show what you want with:
```scala
object CPUTypes {
  object AluMux1Sel extends Chisel Enum { /** */ }
}
```

Reviewer 2: Additionally, it would be good to make clear what the `"selectRS1" -> 0.U` portion is supposed to be doing and what that does. It would be fine to show this with two different variants of `AluMux1Sel`, e.g., `AluMux1Sel` and `AluMex1SelExplicitValues` or something.

Author: Right so this the mapping comment comes from a misunderstanding I had with the value assignment/verification code in the example StrongEnum code. This is the actual code that I have in my file

```scala
object AluMux1Sel extends ChiselEnum {
        val rs1out, pcout = Value
        val correct_annotation_map = Map[String, UInt](elems =
            "rs1out" -> 0.U,
            "pcout"  -> 1.U
        )
    }
```

When I first saw this mapping function, I thought the mapping is what actually assigned the value BUT IT DOESN'T!!! For the documentation I removed the actual function because it is only for verification and not for assigning value.

Reviewer: i think the current form of the doc is helpful, I found it clear what it was saying

Reviewer 2: Ah, I get it now. Maybe just extend the comment to say: "This is what the mapping will be" or something. Not blocking. I'm fine with it the way it is. :+1:

Here we see a mux using the AluMux1Sel to select between different inputs. 

```scala mdoc
import AluMux1Sel._

class AluMux1Bundle extends Bundle {
        val aluMux1Sel =  Input( AluMux1Sel() )
        val rs1Out     =  Input(Bits(32.W))
        val pcOut      =  Input(Bits(32.W))
        val aluMux1Out = Output(Bits(32.W))
}

class AluMux1File extends Module {
    val io = IO(new AluMux1Bundle)

    // Default value for aluMux1Out
    io.aluMux1Out := 0.U

    switch (io.aluMux1Sel) {
        is (selectRS1) {
            io.aluMux1Out  := io.rs1Out
        }
        is (selectPC) {
            io.aluMux1Out  := io.pcOut
        }
    }
}
```
```scala mdoc:verilog
ChiselStage.emitVerilog(new AluMux1File )
```

ChiselEnum also allows for the user to define variables by passing in the value shown below. Note that the value must be increasing or else 

 > chisel3.internal.ChiselException: Exception thrown when elaborating ChiselGeneratorAnnotation

is thrown during Verilog generation.

> ```scala 

Reviewer: ```suggestion
```scala mdoc
```

object Opcode extends ChiselEnum {
    val load  = Value(0x03.U) // i load  -> 000_0011
    val imm   = Value(0x13.U) // i imm   -> 001_0011
    val auipc = Value(0x17.U) // u auipc -> 001_0111
    val store = Value(0x23.U) // s store -> 010_0011
    val reg   = Value(0x33.U) // r reg   -> 011_0011
    val lui   = Value(0x37.U) // u lui   -> 011_0111
    val br    = Value(0x63.U) // b br    -> 110_0011
    val jalr  = Value(0x67.U) // i jalr  -> 110_0111
    val jal   = Value(0x6F.U) // j jal   -> 110_1111
}

> ```

Reviewer 2: Any code block should use the `scala mdoc` tag. This this to get type checked by the compiler and ensure that our documentation isn't heinously busted. 😄

The user can 'jump' to a value and continue incrementing by passing a start point then using a regular Value assignment. 

> ```scala 

Reviewer 2: ```suggestion
```scala mdoc
```

object BranchFunct3 extends ChiselEnum {
    val beq, bne = Value
    val blt = Value(4.U)
    val bge, bltu, bgeu = Value
    /**
        "beq"  -> 0.U,
        "bne"  -> 1.U,
        "blt"  -> 4.U,
        "bge"  -> 5.U,
        "bltu" -> 6.U,
        "bgeu" -> 7.U
    */
}
```

## Testing

When testing your modules, the `.Type` and `.litValue` attributes allow for the the objects to be passed as parameters and for the value to be converted to BigInt type. Note that BigInts cannot be casted to Int with `.asInstanceOf[Int]`, they use their own methods like `toInt`. Please review the [scala.math.BigInt](https://www.scala-lang.org/api/2.12.5/scala/math/BigInt.html) page for more details!

> ```scala 

Reviewer 2: ```suggestion
```scala mdoc
```

def expectedSel(sel: AluMux1Sel.Type): Boolean = sel match {
  case AluMux1Sel.selectRS1 => (sel.litValue == 0)
  case AluMux1Sel.selectPC  => (sel.litValue == 1)
  case _                 => false
}
```

The ChiselEnum type also has methods `.all` and `.getWidth` where `all` returns all of the enum instances and `getWidth` returns the width of the hardware type.

## Workarounds

As of 2/26/2021, the width of the values is always inferred. To work around this, you can add an extra `Value` that forces the width that is desired. This is shown in the example below, where we add a field `ukn` to force the width to be 3 bits wide: 

> ```scala

Reviewer 2: ```suggestion
```scala mdoc
```

object StoreFunct3 extends ChiselEnum {
    val sb, sh, sw = Value
    val ukn = Value(7.U)
    /**
        "sb" -> 0.U,
        "sh" -> 1.U,
        "sw" -> 2.U
    */
}
```

Signed values are not supported so if you want the value signed, you must cast the UInt with `.asSInt`.

## Additional Resources

> The ChiselEnum type is much more powerful than stated above. It allows for Sequence, Vec, and Bundle assignments as well as the `.next` operation to allow for stepping through sequential states and the `.isValid` for checking that the value is an enum type. The source code for the ChiselEnum can be found [here](https://github.com/chipsalliance/chisel3/blob/2a96767097264eade18ff26e1d8bce192383a190/core/src/main/scala/chisel3/StrongEnum.scala) in the class `EnumFactory`. An example file for many of the ChiselEnum operations can be found [here](https://github.com/chipsalliance/chisel3/blob/dd6871b8b3f2619178c2a333d9d6083805d99e16/src/test/scala/chiselTests/StrongEnum.scala).

Reviewer: ```suggestion
The ChiselEnum type is much more powerful than stated above. It allows for Sequence, Vec, and Bundle assignments, as well as a `.next` operation to allow for stepping through sequential states and an `.isValid` for checking that a hardware value is a valid `Value`. The source code for the ChiselEnum can be found [here](https://github.com/chipsalliance/chisel3/blob/2a96767097264eade18ff26e1d8bce192383a190/core/src/main/scala/chisel3/StrongEnum.scala) in the class `EnumFactory`. Examples of the ChiselEnum operations can be found [here](https://github.com/chipsalliance/chisel3/blob/dd6871b8b3f2619178c2a333d9d6083805d99e16/src/test/scala/chiselTests/StrongEnum.scala).
```

// package CPUTypes {
    object AluMux1Sel extends ChiselEnum {
        val selectRS1, selectPC = Value

>         /** How the values will be mapped

Reviewer 2: :100:
